add function to create expected outputs for deterministic tests

ansalma6 opened the pull request:
PR for https://github.com/Lind-Project/lind-wasm/issues/11

Added a function in wasmtest.sh file for creating expected outputs. 
The function can be ran using "wasmtest.sh createexpected" or "wasmtest.sh ce"

All c files under each of the deterministic subfolder are taken and compiled using gcc. 
A new folder expected is created in each deterministic subfolder and the output is saved in a <test_case_filename>.output file
If there are any compile or runtime error, they are displayed in the console and alse added to the .output file.
The compiled files are deleted after running
Finally a count of total/success/failure is displayed in the console

m-hemmings commented on wasmtest.sh:
```diff
@@ -198,6 +198,54 @@ run_from_files() {
+    total_success=0
+    total_failure=0
+
+    for deterministic_dir in $(find tests/unit-tests -type d -name deterministic); do
```
Instead of hard coding the name for the file to be found, could it be added as a parameter or configurable variable?
Something like:
`TARGET_DIR_NAME="deterministic"
for deterministic_dir in $(find tests/unit-tests -type d -name "$TARGET_DIR_NAME"); do`

ansalma6 replied:
Changed to use variable

m-hemmings commented on wasmtest.sh:
```diff
@@ -198,6 +198,54 @@ run_from_files() {
     echo -e "${GREEN}All tests from file list completed.${RESET}"
 }
 
+#function to create the expected outputs 
```
Could the comments at the head of the function be clearer as to the purpose and flow of the function? Perhaps something like:
```
# ----------------------------------------------------------------------
# Function: create_expected_outputs
#
# Purpose:
# Compiles `.c` files in "tests/unit-tests/deterministic" directories, 
# runs the binaries, and saves their output in "expected" subdirectories.
#
# Variables:
# - Input: None required; works on `.c` files in specific directories.
# - Output: Logs total processed, successes, and failures to stdout.
#
# Exceptions:
# - Compilation errors: Logged to stdout and corresponding `.output` files.
# - Runtime errors: Logged to stdout and corresponding `.output` files.
# - Skips directories or files if not found; no explicit errors thrown.
#
# Note:
# Temporarily disables `set -e` and `set -o pipefail` for error handling.
# ----------------------------------------------------------------------

```

ansalma6 replied:
Added the header

Yaxuan-w reviewed (commented):
I was wondering if we might consider adding a cleanup as an optional feature as well. Since we will create a subfolder for each deterministic category, manually cleaning them up could become messy. 

Would it be possible to include something like `wasmtest.sh cleanup` or `wasmtest.sh cl`, to remove all the `/expected` folders and their contents?

ansalma6 commented:
Changes from Review comments:
Added a new function to cleanup the generated expected outputs "wasmtest.sh cleanexpected" or "wasmtest.sh cle"
Added comment headers for functions
Outputs the paths of successful tests to testlist_safe.txt and failed tests to testlist_unsafe (one file per line)

Note:
Renamed the function to generate expected outputs to avoid potential confusion with cleanup arguments.
"wasmtest.sh generateexpected" or "wasmtest.sh ge"

rennergade reviewed (approved):
This looks good, agree with the previous comments. @yashaswi2000 We should probably try to cache these expected outputs though since they shouldnt change.

Yaxuan-w reviewed (approved):
LGTM

yashaswi2000 reviewed (commented):
LGTM!
